apm-agent-plugins/apm-opentracing-plugin/src/main/java/co/elastic/apm/opentracing/impl/ApmSpanInstrumentation.java

```diff
@@ -74,34 +73,48 @@ public FinishInstrumentation() {
             super(named("finishInternal"));
         }
 
+        @Advice.OnMethodEnter
+        private static void finishInternal(@Advice.FieldValue(value = "dispatcher", readOnly = false) @Nullable Object dispatcher,
+                                           @Advice.Argument(0) long finishMicros) {
+            final Object newDispatcher = doFinishInternal(dispatcher, finishMicros);
+            if (newDispatcher != null) {
+                dispatcher = newDispatcher;
```

Comment:
I still think this doesn't solve the async problem. Consider this scenario:
Thread A starts a transaction, then submits a task to a queue, then the task starts being executed on thread B (original transaction is still running on thread A). Now thread B wants to do something, for example - set the operation name:
```
Thread B:
----------
    apmTracer.activeSpan().setOperationName("opName");
        ApmSpanInstrumentation$SetOperationName.setOperationName()
            if (dispatcher instanceof AbstractSpan) {     // returns true
Thread A:
----------
    apmTracer.activeSpan().finish();
        ApmSpanInstrumentation.FinishInstrumentation.finishInternal()
            ApmSpanInstrumentation.FinishInstrumentation.doFinishInternal()
                span.end()     // potentially recycled by thread A or even thread C (disruptor thread)
        dispatcher = newDispatcher;    // this line - replace the underlying dispatcher from span to serialized context
Thread B:
----------
            // continue execution
            AbstractSpan<?> span = (AbstractSpan) dispatcher;    // still the reference to the span
            span.setName(operationName);    // on a potentially recycled span
```

or this scenario:
```
Thread A:
----------
    apmTracer.activeSpan().finish();
        ApmSpanInstrumentation.FinishInstrumentation.finishInternal()
            ApmSpanInstrumentation.FinishInstrumentation.doFinishInternal()
                span.end()     // potentially recycled by thread A or even thread C (disruptor thread)
Thread B:
----------
    apmTracer.activeSpan().setOperationName("opName");
        ApmSpanInstrumentation$SetOperationName.setOperationName()
            if (dispatcher instanceof AbstractSpan) {   // returns true as the dispatcher ref was not yet replaced
                AbstractSpan<?> span = (AbstractSpan) dispatcher;
                span.setName(operationName);    // on a potentially recycled span
Thread A:
----------
        // continue execution
        dispatcher = newDispatcher;    // this line - only now replacing the underlying dispatcher, too late
```